In [1]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd
from scipy import stats

# Part 1: Load the dataset

In [2]:
# Load the given datasets
kidney_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
kidney_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print(kidney_categorical.head())
print(kidney_numerical.head())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
This data shows various attributes about a person's health and whether they have chronic kidney disease (CKD) or not.
There are 24 different features the data considers including age, blood pressure, and whether the person has an appetite.
There is only 1 label: whether the person has CKD or not.

This dataset contains both categorical and numerical data.  Pus Cell Clumps, appetite, and anemia are all categorical variables.  Age, blood pressure, and sodium levels are all numerical variables.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

kidney_numerical_dupe = kidney_numerical.duplicated(subset=['unique_id'])
print("# duplicated rows (numerical)", kidney_numerical_dupe.to_list().count(True))

# Drop these duplicate rows
for i, dupe in enumerate(kidney_numerical_dupe):
    if dupe:
        kidney_numerical.drop(i, inplace=True)

print("rows left:", len(kidney_numerical))

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
kidney_categorical_dupe = kidney_categorical.duplicated(subset=['unique_id'])
print("# duplicated rows (categorical)", kidney_categorical_dupe.to_list().count(True))

for i, dupe in enumerate(kidney_categorical_dupe):
    if dupe:
        kidney_categorical.drop(i, inplace=True)

print("rows left:", len(kidney_categorical))

# duplicated rows (numerical) 30
rows left: 390
# duplicated rows (categorical) 30
rows left: 400


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
All unique ids are not present in both datasets, since they have different lengths. Numerical has 390 items while categorical has 400 items. For rows that are present in one dataset but missing in the other will have corresponding empty values in the combined table.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
kidney_df = pd.merge(kidney_categorical, kidney_numerical, how="outer", on="unique_id")

#Print the combined dataset
kidney_df

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,...,60.0,125.0,21.0,1.3,137.0,3.4,15.0,46.0,NaN,NaN
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,...,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,...,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
396,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9
397,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,...,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8
398,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,117.0,22.0,1.2,138.0,3.5,13.0,45.0,5200.0,5.6


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
num_missing = len(kidney_df[kidney_df.isna().sum(axis=1) > 0])

# Print %
print(f"Rows missing values: {100*num_missing/len(kidney_df)}%")

# Drop these rows from the dataset
kidney_df_dropped = kidney_df.dropna(axis=0, how="any")

# Print the Dataset
kidney_df_dropped.head()

Rows missing values: 61.75%


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,...,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3


## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
kidney_df_sorted = kidney_df_dropped.sort_values("Target", ignore_index=True)

# Print the dataset
kidney_df_sorted.head()


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,534245,2.0,0.0,normal,abnormal,present,notpresent,no,no,no,...,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,19100.0,3.7
1,349892,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,...,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1
2,343710,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,...,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2
3,102677,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1
4,514721,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

kidney_df_encoded = pd.get_dummies(kidney_df_sorted, categorical_columns,"_")

# Print the dataset
kidney_df_encoded

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,534245,2.0,0.0,45.0,70.0,117.0,52.0,2.2,136.0,3.8,...,1,0,1,0,1,0,1,0,1,0
1,349892,1.0,3.0,59.0,70.0,424.0,55.0,1.7,138.0,4.5,...,0,1,1,0,1,0,1,0,1,0
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,...,0,1,0,1,0,1,0,1,1,0
3,102677,3.0,0.0,59.0,70.0,76.0,186.0,15.0,135.0,7.6,...,1,0,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

For binary variables, storing the value in a single column does work. In this representation, each extreme (0 or 1) represents whether the binary variable is true or false. This would be equivalent since var_no is always the complement of var_yes, meaning no extra information is gained from the extra column.

If the categorical variable could take mor ethan 2 values, then representing it using a single column would not work. These categorical values would all be on the same continuous spectrum, which poses a problem. If the condiments ranch and ketchup have assigned values 1 and 5, the input essentially reads ketchup as a more extreme version of ranch.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [25]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']


kidney_df_no_outliers = kidney_df_encoded[(np.abs(stats.zscore(kidney_df_encoded[numerical_columns], axis=0)) < 3).all(axis=1)]
# Print the dataset
kidney_df_no_outliers


,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,...,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,...,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,69.0,70.0,264.0,87.0,2.7,130.0,4.0,...,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,62.0,70.0,122.0,42.0,1.7,136.0,4.7,...,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,6.0,60.0,94.0,67.0,1.0,135.0,4.9,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [30]:
# Normalize the all Numerical Attributes in the dataset.
kidney_df_norm = kidney_df_no_outliers.copy()
for col in numerical_columns:
    min_val = np.min(kidney_df_no_outliers[col])
    max_val = np.max(kidney_df_no_outliers[col])
    kidney_df_norm[col] = (kidney_df_no_outliers[col] - min_val)/(max_val-min_val) #normalize 0 to 1

# Print the dataset
kidney_df_norm

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,...,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,...,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,...,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,...,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.600000,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.314286,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.285714,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.285714,...,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [35]:
#Remove that column
kidney_df_final = kidney_df_norm.drop(labels="unique_id", axis=1)
# Print the dataset
kidney_df_final

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,...,0,1,0,1,0,1,0,1,1,0
4,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,...,1,0,1,0,1,0,1,0,1,0
7,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,...,0,1,1,0,0,1,1,0,1,0
8,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,...,1,0,1,0,1,0,1,0,1,0
10,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,0.352459,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.0,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.600000,0.983607,...,1,0,1,0,1,0,1,0,0,1
149,0.0,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.314286,0.713115,...,1,0,1,0,1,0,1,0,0,1
150,0.0,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.285714,0.745902,...,1,0,1,0,1,0,1,0,0,1
151,0.0,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.285714,0.614754,...,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [45]:
# Export the dataframe to a new csv file
kidney_df_final.to_csv('./kidney_results.csv', index=False)
# Export the dataframe to a new json file
kidney_df_final.to_json('./kidney_results.json')
# no pickle??

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

Note: this overflows a bit to the right

| al   | su   | age                      | bp   | bgr                     | bu                        | sc                      | sod                     | pot                     | hemo                    | pcv                     | wbcc                    | rbcc                    | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|-------------------------|------|-------------------------|--------------------------|-------------------------|-------------------------|-------------------------|-------------------------|-------------------------|-------------------------|-------------------------|--------------|------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|------------|----------------|
| 2.0  | 0.0  | 0.7432432432432432      | 0.5  | 0.44206008583690987    | 0.9019607843137255       | 0.4320987654320988      | 0.5                     | 0.6571428571428571      | 0.17213114754098363     | 0.21052631578947367     | 0.3951612903225806      | 0.15384615384615388     | 1            | 0          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0          | 1          | 0     | 1      | 0      | 1       | 0          | 1              |
| 2.0  | 0.0  | 0.6756756756756757      | 0.75 | 0.2532188841201717     | 0.6339869281045751      | 0.7777777777777778      | 0.36666666666666664     | 0.5428571428571428      | 0.28688524590163933     | 0.34210526315789475     | 0.1693548387096774      | 0.2051282051282051      | 1            | 0          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 1     | 0      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 3.0  | 4.0  | 0.8513513513513513      | 0.25 | 0.8326180257510729     | 0.5032679738562091      | 0.28395061728395066     | 0.3333333333333333      | 0.3142857142857143      | 0.5655737704918032      | 0.5526315789473685      | 0.4274193548387097      | 0.3846153846153845      | 0            | 1          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 0          | 1          | 1     | 0      | 1      | 0       | 1          | 0              |
| 3.0  | 0.0  | 0.7567567567567568      | 0.25 | 0.22317596566523606    | 0.20915032679738563     | 0.16049382716049382     | 0.5333333333333333      | 0.5142857142857143      | 0.5737704918032787      | 0.6052631578947368      | 0.2903225806451613      | 0.3333333333333333      | 0            | 1          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 4.0  | 0.0  | 0.0                     | 0.0  | 0.10300429184549356    | 0.37254901960784315     | 0.07407407407407407     | 0.5                     | 0.5714285714285715      | 0.3524590163934426      | 0.3684210526315789      | 1.0                     | 0.5641025641025641      | 1            | 0          | 1           | 0         | 1              | 0           | 0             | 1          | 1      | 0       | 1     | 0      | 1      | 0       | 0          | 1          | 1     | 0      | 1      | 0       | 1          | 0              |
| 3.0  | 1.0  | 0.6621621621621622      | 0.5  | 0.6180257510729614     | 0.4117647058823529      | 0.4320987654320988      | 0.5666666666666667      | 0.5714285714285715      | 0.43442622950819676     | 0.47368421052631576     | 0.25                    | 0.2820512820512821      | 0            | 1          | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |
| 1.0  | 0.0  | 0.5405405405405406      | 0.0  | 0.39914163090128757    | 0.5359477124183006      | 0.35802469135802467     | 0.7                     | 0.3142857142857143      | 0.34426229508196726     | 0.3157894736842105      | 0.8306451612903226      | 0.15384615384615388     | 0            | 1          | 0           | 1         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 4.0  | 2.0  | 0.7837837837837838      | 1.0  | 0.39914163090128757    | 0.2875816993464052      | 0.8395061728395062      | 0.6666666666666666      | 0.4857142857142856      | 0.18852459016393447     | 0.2631578947368421      | 0.25806451612903225     | 0.2051282051282051      | 1            | 0          | 1           | 0         | 1              | 0           | 0             | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 1          | 0          | 1     | 0      | 0      | 1       | 1          | 0              |
| 4.0  | 0.0  | 0.5675675675675675      | 0.5  | 0.1072961373390558     | 1.0                      | 0.9012345679012346      | 0.5333333333333333      | 0.25714285714285706     | 0.34426229508196726     | 0.42105263157894735     | 0.20967741935483872     | 0.2051282051282051      | 0            | 1          | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 1     | 0      | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |
| 4.0  | 3.0  | 0.8513513513513513      | 0.25 | 0.6180257510729614     | 0.5620915032679739      | 0.7283950617283951      | 0.0                     | 0.2857142857142857      | 0.3114754098360656      | 0.3157894736842105      | 0.5806451612903226      | 0.17948717948717943     | 0            | 1          | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 1          | 0          | 0     | 1      | 0      | 1       | 1          | 0              |
| 3.0  | 2.0  | 0.9054054054054054      | 1.0  | 0.9656652360515021     | 0.5228758169934641      | 0.6419753086419753      | 0.6666666666666666      | 0.0                     | 0.2950819672131147      | 0.3684210526315789      | 0.21774193548387097     | 0.15384615384615388     | 1            | 0          | 1           | 0         | 0              | 1           | 0             | 0          | 1      | 0       | 0     | 1      | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 4.0  | 0.0  | 0.20270270270270271     | 0.75 | 0.15879828326180256    | 0.19607843137254902     | 0.16049382716049382     | 0.16666666666666666     | 0.17142857142857143     | 0.22131147540983614     | 0.18421052631578946     | 0.6532258064516129      | 0.3333333333333333      | 0            | 1          | 1           | 0         | 0              | 1           | 0             | 1          | 1      | 0       | 1     | 0      | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |
| 4.0  | 1.0  | 0.7837837837837838      | 0.0  | 0.7253218884120172     | 0.3137254901960784      | 0.4814814814814815      | 0.5666666666666667      | 0.7142857142857143      | 0.319672131147541       | 0.34210526315789475     | 0.25806451612903225     | 0.2051282051282051      | 1            | 0          | 1           | 0         | 1              | 0           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1          | 0          | 0     | 1      | 0      | 1       | 1          | 0              |
| 4.0  | 1.0  | 0.6756756756756757      | 0.25 | 0.6008583690987125     | 0.10457516339869281     | 0.16049382716049382     | 0.5333333333333333      | 0.25714285714285706     | 0.8606557377049181      | 0.9473684210526315      | 0.6612903225806451      | 0.7692307692307692      | 1            | 0          | 0           | 1         | 1              | 0           | 1             | 0          | 1      | 0       | 1     | 0      | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 4.0  | 0.0  | 0.5675675675675675      | 0.5  | 0.2703862660944206     | 0.8431372549019608      | 1.0                     | 0.4                     | 0.7428571428571428      | 0.3852459016393443      | 0.5263157894736842      | 0.1532258064516129      | 0.358974358974359       | 1            | 0          | 1           | 0         | 1              | 0           | 0             | 1          | 1      | 0       | 0     | 1      | 1      | 0       | 1          | 0          | 1     | 0      | 0      | 1       | 1          | 0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

We could potentially split the dataset into smaller batches for GPT to process, so each individual batch has a lower chance to hallucinate.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [42]:
# Code to rename all the columns in the dataset

column_mapping = {
    'age': 'age (years)',
    'bp': 'blood pressure (mm/Hg)',
    'sc': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc_abnormal': 'red blood cells abnormal',
    'rbc_normal': 'red blood cells normal',
    'pc_abnormal': 'pus cell abnormal',
    'pc_normal': 'pus cell normal',
    'pcc_notpresent': 'pus cell clumps not present',
    'pcc_present': 'pus cell clumps present',
    'ba_notpresent': 'bacteria not present',
    'ba_present': 'bacteria present',
    'bgr': 'blood glucose random (mgs/dl)',
    'bu': 'blood urea (mgs/dl)',
    'sod': 'sodium (mEq/L)',
    'pot': 'potassium (mEq/L)',
    'hemo': 'hemoglobin (gms)',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count (cells/cumm)',
    'rbcc': 'red blood cell count (millions/cmm)',
    'htn_no': 'hypertension no',
    'htn_yes': 'hypertension yes',
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'class ckd',
    'Target_notckd': 'class notckd',
}

# Rename the DataFrame columns
kidney_df_final.rename(columns=column_mapping, inplace=True)

kidney_df_final

,albumin,sugar,age (years),blood pressure (mm/Hg),blood glucose random (mgs/dl),blood urea (mgs/dl),specific gravity,sodium (mEq/L),potassium (mEq/L),hemoglobin (gms),...,coronary artery disease no,coronary artery disease yes,appetite good,appetite poor,pedal edema no,pedal edema yes,anemia no,anemia yes,class ckd,class notckd
2,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,...,0,1,0,1,0,1,0,1,1,0
4,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,...,1,0,1,0,1,0,1,0,1,0
7,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,...,0,1,1,0,0,1,1,0,1,0
8,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,...,1,0,1,0,1,0,1,0,1,0
10,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,0.352459,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.0,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.600000,0.983607,...,1,0,1,0,1,0,1,0,0,1
149,0.0,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.314286,0.713115,...,1,0,1,0,1,0,1,0,0,1
150,0.0,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.285714,0.745902,...,1,0,1,0,1,0,1,0,0,1
151,0.0,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.285714,0.614754,...,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [47]:
# Converted SQL to Pandas code
# Group by 'Target' and count the occurrences
result = kidney_df_final.groupby('class ckd').size().reset_index(name='count')

# Display the result
print(result)

   class ckd  count
0          0    111
1          1     23


The GPT-generated code provides a total count of patients with CKD and without CKD.